### PR TITLE
Implement `LIKE` for non-string, non-literals

### PR DESCRIPTION
**Pull request by alancai98**
(Additional LIKE tests are in pending partiql-tests PR: https://github.com/partiql/partiql-tests/pull/49. Once that's merged, will update the partiql-tests submodule).

Implements `LIKE` for non-string, non-literals. Previous implementation only accounted for scenario in which `LIKE` was given non-string, non-literal arguments. This resulted in some queries given NULL or MISSING pattern or escape arguments to fail an assertion while lowering to the logical plan.

Remaining TODOs:
- [x]  Documentation + cleanup
- [x] Sorting out if `LIKE` follows 3 or 4 value logic
- [x] Add some edge cases missing from the conformance tests (pending merging of https://github.com/partiql/partiql-tests/pull/49) -- this should also fulfill the codecoverage requirements

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-eval/src/eval.rs [hunk 1]
```diff
@@ -1155,7 +1156,58 @@ impl EvalExpr for EvalLikeMatch {
...
+    ) -> Self {
+        EvalLikeNonStringNonLiteralMatch {
+            value,
+            pattern,
+            escape,
+        }
+    }
+}
+
+impl EvalExpr for EvalLikeNonStringNonLiteralMatch {
```
Comment (alancai98) [inline partiql-eval/src/eval.rs:1187]: Added another node so the simple case of a string literal pattern (and string literal escape) (e.g. `col LIKE 'fo%`) wouldn't have to be converted to regex for every `value`

File: partiql-conformance-tests/src/bin/generate_comparison_report.rs [hunk 2]
```diff
@@ -135,7 +135,7 @@ Number failing in main but now pass: {}
         ).expect("write passing_orig_failing_new heading");
         for test_name in &passing_orig_failing_new {
             comparison_report_file
-                .write_all(format!("- {}\n", test_name).as_bytes())
+                .write_all(format!("- {test_name}\n").as_bytes())
```
Comment (alancai98) [inline partiql-conformance-tests/src/bin/generate_comparison_report.rs:138]: These changes and from `generate_cts_report.rs` were other clippy warnings not found fixed in #288.

File: partiql-logical-planner/src/lower.rs [hunk 3]
```diff
@@ -737,18 +722,52 @@ impl<'ast> Visitor<'ast> for AstToLogical {
...
+                    Pattern::LikeNonStringNonLiteral(LikeNonStringNonLiteralMatch {
+                        pattern: Box::new(pattern_ve),
+                        escape: Box::new(escape_ve),
+                    });
+                self.push_vexpr(ValueExpr::PatternMatchExpr(PatternMatchExpr {
+                    value,
+                    pattern: pattern_non_string_literals,
+                }))
+            }
+        }
```
Comment (jpschorr) [inline partiql-logical-planner/src/lower.rs:770]: Perhaps pull out the `push_vexpr`. I think it makes it a bit clearer what the logic is.

```suggestion
      let pattern = match (&pattern_ve, &escape_ve) {
          (ValueExpr::Lit(pattern_lit), ValueExpr::Lit(escape_lit)) => {
              match (pattern_lit.as_ref(), escape_lit.as_ref()) {
                  (Value::String(pattern), Value::String(escape)) => Pattern::Like(LikeMatch {
                      pattern: pattern.to_string(),
                      escape: escape.to_string(),
                  }),
                  _ => Pattern::LikeNonStringNonLiteral(LikeNonStringNonLiteralMatch {
                      pattern: Box::new(pattern_ve),
                      escape: Box::new(escape_ve),
                  }),
              }
          }
          _ => Pattern::LikeNonStringNonLiteral(LikeNonStringNonLiteralMatch {
              pattern: Box::new(pattern_ve),
              escape: Box::new(escape_ve),
          }),
      };

      let pattern = ValueExpr::PatternMatchExpr(PatternMatchExpr { value, pattern });
      self.push_vexpr(pattern);
```

**Review by jpschorr — APPROVED** [review]

File: partiql-logical-planner/src/lower.rs (hunk 3 above)
Comment (alancai98) [reply partiql-logical-planner/src/lower.rs:770]: Thanks for the suggestion. Moving the `push_vexpr` out of the match expression made things a lot clearer.

**Review by am357 — APPROVED** [review]
